File: bytopic/binarysearch/java/group1/SearchInsertPosition.java
// Given a sorted array of distinct integers and a target value, return the index if the target is found.
// If not, return the index where it would be if it were inserted in order.
// You must write an algorithm with O(log n) runtime complexity.
// Example 1:
// Input: nums = [1,3,5,6], target = 5
// Output: 2
// Example 2:
// Input: nums = [1,3,5,6], target = 2
// Output: 1
//
// Tag: 114/150
// Tag: 35/2927, R434/2936 (overall frequency ranking)

class SearchInsertPosition {
    public static void main(String[] args) {
        int[] nums = {1,3,5,6};
        int target = 2;
        System.out.println("Search insert position: " + searchInsert(nums, target));
    }

    static int searchInsert(int[] nums, int target) {
        int start = 0;
        int end = nums.length - 1;

        while (start <= end) {
            int mid = start + (end - start) / 2;
            if (nums[mid] == target)
                return mid;
            else if (nums[mid] > target)
                end = mid - 1;
            else start = mid + 1;
        }

        return start;
    }
}

// Time complexity ~ O(logn)
// Space Complexity ~ O(1)

File: bytopic/binarysearch/python/group1/search_insert_position.py
"""
Given a sorted array of distinct integers and a target value, return the index if the target is found.
If not, return the index where it would be if it were inserted in order.
You must write an algorithm with O(log n) runtime complexity.

Example 1:
Input: nums = [1,3,5,6], target = 5
Output: 2

Example 2:
Input: nums = [1,3,5,6], target = 2
Output: 1

Tag: 114/150
Tag: 35/2927, R434/2936 (overall frequency ranking)
"""


def search_insert(nums: list[int], target: int) -> int:
    left = 0
    right = len(nums) - 1
    while right >= left:
        mid = (left + right) // 2
        if nums[mid] == target:
            return mid
        elif nums[mid] < target:
            left = mid + 1
        else:
            right = mid -1
    return left


def main():
    nums = [1, 3, 5, 6]
    target = 5
    print(search_insert(nums, target))


if __name__ == "__main__":
    main()
